resource_customizations/kubevirt.io/VirtualMachine/health.lua

```diff
@@ -0,0 +1,15 @@
+hs = { status="Progressing", message="No status available"}
```

Comment:
The VM state is computed by its controller and is derived from quite a few factors. The VM's `.status.printableStatus` field captures a human-readable summary of this state. We can use the value of this field in the `hs.message`, e.g. `The VirtualMachine is in the '%s' state", obj.status.printableStatus`.

As for the `hs.status`, I think the following is a reasonable mapping to Argo CD [health states](https://github.com/argoproj/gitops-engine/blob/master/pkg/health/health.go):

- !created --> Suspended
- created && Paused --> Suspended
- created && Ready --> Healthy
- created && !Ready --> Progressing

As for the last one, it isn't always accurate since some states better map to Degrading (e.g., various error states), but printableStatus in the message can be used to convey that extra information.

Reply:
For every permutation of status and message,  I will need to write a test with an example manifest.
My original thought was to follow the `.status.conditions` but this only updates IFF the VM is running.

I like the idea of providing some outputs to Suspended, as that could provide useful information where manual intervention is needed to get an application to progress. I will experiment with this more.

Reply:
The  suggestion looks great, one point in addition:

> * created && !Ready --> Progressing

If a VM would be come unhealthy it would go back to `Progressing` instead of `Degraded`. I am not sure if one can detect right now immediately on the status if the VM was ready before and became unready.

One way could work like this:

 * We safe the phase transition timestamps on the VMIs. One could look up the initial start delay for readiness probes, take the running transition timestamp and show "progressing" until timestamp + readiness delay is over and from that timepoint on always show "degraded".

Reply:
> I am not sure if one can detect right now immediately on the status if the VM was ready before and became unready.

You're right, in this case the VM's printableStatus would still show Running, and alongside a Progressing argocd status - we're missing an adequate indication that something's wrong with the VM.

> We safe the phase transition timestamps on the VMIs

The context in which this lua script operates, have access only to the VM object, and can't (and, IMO, shouldn't) query the kubernetes API to fetch the corresponding VMI.

Another option is to look at the VM's printableStatus, and if it's Running and !Ready, then set a Degraded status. Can also do that for a Migrating (&& !Ready) printableStatus.
True, this violates our intention of not to write automation based on the printableStatus since its states might be changed/added/removed. But if we consider that the Running state is not likely to ever disappear, this can probably present more value than risk.

Reply:
I just pushed updates to the VM health check. I center on the status.printableStatus field, with a simplistic fallback in case of an old version missing the field.

Reply:
Hm, `status.printableStatus` is a human-readable only field. It can change anytime and should not be relied on.

Reply:
If I am not mistaken we already mirror the readiness probe to the VM too. That should be enough to find out about degraded states.

Reply:
@rmohr I've made an adjustment here to set Healthy based on `status.ready` instead of printableStatus.

The other states default to Progressing and printableStatus, so I think we are okay even if the style of printableStatus changes. The only status we want to get 100% correct is Healthy.